pkg/container/SPEC.md

```diff
@@ -0,0 +1,103 @@
+{
+  "ctn-v1": [
+    <token1 bytes>,
+    <token2 bytes>,
+    <token3 bytes>,
+  ]
+}
+```
+
+## 2.2 Serialisation
+
+To serialize the container into bytes, the inner CBOR structure MUST be serialized into bytes according to the CBOR specification. The resulting bytes MAY be compressed by a supported algorithm, then MAY be encoded with a supported base encoding.
+
+The following compression algorithm are REQUIRED to be supported:
+- [GZIP]
+
+The following base encoding combination are REQUIRED to be supported:
+- base64, standard alphabet, padding
+- base64, URL alphabet, no padding
+
+The CBOR bytes MUST be prepended by a single byte header to indicate the selection combination of base encoding and compression. This header value MUST be set according to the following table:
+
+| Header as hex | Header as ASCII | Base encoding            | Compression    |
+|---------------|-----------------|--------------------------|----------------|
+| 0x40          | @               | raw bytes                | no compression | 
+| 0x42          | B               | base64 std padding       | no compression | 
+| 0x43          | C               | base64 url               | no compression | 
+| 0x4D          | M               | raw bytes                | gzip           | 
+| 0x4F          | O               | base64 std padding       | gzip           | 
+| 0x50          | P               | base64 url               | gzip           | 
+
+# 3 FAQ
+
+## 3.1 Why not include the UCAN CIDs?
+
+Several attacks are possible if UCAN tokens aren't validated. If CIDs aren't validated, at least two attacks are possible: [privilege escalation] and [cache poisoning], as UCAN delegation proofs depends on a correct hash-linked structure.
+
+By not including the CID in the container, the recipient is forced to hash (and thus validate) the CIDs for each token. If presented with a claimed CID paired with the token bytes, implementers could ignore CID validation, breaking a core part of the proof chain security model. Hash functions are very fast on a couple kilobytes of data so the overhead is still very low. It also reduces significantly the size of the container.
+
+## 3.2 Why compress? Why not always compress?
+
+Compression is a relatively demanding operation. As such, using it is a tradeoff between size on the wire and CPU/memory usage, both when writing and reading a container. The transport itself can make compression worthwhile or note: for example, HTTP/2 and HTTP/3 headers are already compressed, but HTTP/1 headers are not. This being highly contextual, the choice is left to the final implementer.
+
+# 4 Implementation recommendations
+
+## 4.1 Dissociate reader and writer
+
+While it tempting to write a single implementation to read and write a container, it is RECOMMENDED to separate the implementation into a reader and a writer. The writer can simply accept arbitrary tokens as bytes, while the reader provide a read-only view with convenient access functions.
+
```

Comment:
I'd also recommend to implementer and users to not use the compression here over HTTP since most HTTP connections in 2025 automatically GZIP and zipping multiple times is wasted effort (and may compress worse overall in some cases when other data like HTTP headers getting included in the GZIP)

Reply:
3.2 talks about this specifically